### ISSUE TITLE
User creation error

### ISSUE THREAD

**Issue by bob109** [hashbang/hashbang.sh#19]
Coming up with a "curl: (22) The requested URL returned error: 400 BAD REQUEST" error trying to create a file.  

Here it is in context. 

 -------------------------------------------------------------------- 
 
 To create your account we first need a username.
 
 A valid username must:
  * be between between 1-31 characters long
  * consist of only 0-9 and a-z (lowercase only)
  * begin with a letter
 
 Traditional unix usernames are first initial, optional middle initial,
 and the first 6 characters of the last name, but feel free to use 
 whatever you want
 
 Username: notbob
 
 -------------------------------------------------------------------- 
 
 Now we will need an SSH Public Key.
 
 SSH Keys are a type of public/private key system that let you identify 
 yourself to systems like this one without ever sending your password 
 over the internet, and thus by nature we won't even know what it is
 
 We found a public key in [ ~/.ssh/id_rsa.pub ]. Use this key? [Y/n] n
 Do you want us to generate a key for you? [Y/n] n
 
 Please enter path to SSH Public Key: 
~/.ssh/id_rsa.hashbang.pub
 
 "~/.ssh/id_rsa.hashbang.pub" does not exist
 
 Please enter path to SSH Public Key: 
/home/bob/.ssh/id_rsa.hashbang.pub
 
 -------------------------------------------------------------------- 
 
 We are going to create an account with the following information
 
 Username: notbob
 Public Key: /home/bob/.ssh/id_rsa.hashbang.pub
 Host: va1
 
 Does this look correct? [Y/n] 
 
 Creating your account...
 
curl: (22) The requested URL returned error: 400 BAD REQUEST
 
 Account creation failed.
 Something went awfully wrong and we couldn't create an account for you.
 If you think this is a bug, please report it to 
 -> https://github.com/hashbang/hashbang.sh/issues/
 
 The installer will not continue from here...
 
The advanced method also fails 
$ curl -d '{"user":"notbob","key":"'"$(cat ~/.ssh/id_rsa.hashbang.pub)"'"}' -H 'Content-Type: application/json' https://hashbang.sh/user/create
{"message": "User creation script failed"}

**Comment by daurnimator** [comment]
> Do you want us to generate a key for you? [Y/n] n
> 
> Please enter path to SSH Public Key: 
> ~/.ssh/id_rsa.hashbang.pub
> 
> "~/.ssh/id_rsa.hashbang.pub" does not exist
> 
> Please enter path to SSH Public Key: 
> /home/bob/.ssh/id_rsa.hashbang.pub

I notice you said 'no' to creating a key, and then specified the same key...
Did you create this file in another terminal?

what does `echo curl -d '{"user":"notbob","key":"'"$(cat ~/.ssh/id_rsa.hashbang.pub)"'"}' -H 'Content-Type: application/json' https://hashbang.sh/user/create` print?

**Comment by bob109** [comment]
I created a key manually (~/.ssh/id_rsa.hashbang.pub).  because I wanted to use this key, I said No don't create one for me and use ~/.ssh/id_rsa.hashbang.pub.  I'm guessing it had problems with the ~ shortcut and the 2nd time I typed in the full path name /home/bob/.ssh/id_rsa.hashbang.pub.  That apparently went though.

Note, I get the same error
curl: (22) The requested URL returned error: 400 BAD REQUEST
When I let it use the ~/.ssh/id_rsa.pub that hashbang.sh found

The advanced method of calling user/create returns
{"message": "User creation script failed"}

Just for grins I also tried to let hashbang.sh create an ssh key for me and get this.  (edited for brevity)

Now we will need an SSH Public Key.
 
 SSH Keys are a type of public/private key system that let you identify 
 yourself to systems like this one without ever sending your password 
 over the internet, and thus by nature we won't even know what it is
 
 We found a public key in [ ~/.ssh/id_rsa.pub ]. Use this key? [Y/n] n
 Do you want us to generate a key for you? [Y/n] y
Generating public/private rsa key pair.
Enter file in which to save the key (/home/bob/.ssh/id_rsa): /home/bob/.ssh/id_rsa_hashbang
Enter passphrase (empty for no passphrase): 
Enter same passphrase again: 
Your identification has been saved in /home/bob/.ssh/id_rsa_hashbang.
Your public key has been saved in /home/bob/.ssh/id_rsa_hashbang.pub.
The key fingerprint is:
e3:99:75:d5:9c:16:9b:18:8d:91:36:c7:d1:bf:79:bf #! notbob
The key's randomart image is:
+--[ RSA 2048]----+
|             o*oo|
|             =o**|
|            ..+=+|
|             .. .|
|        S . .   o|
|       . = .   o.|
|        +       o|
|                .|
|               E.|
+-----------------+
 
 -------------------------------------------------------------------- 
 
 We are going to create an account with the following information
 
 Username: notbob
 Public Key: ~/.ssh/id_rsa.pub
 Host: va1
 
 Does this look correct? [Y/n]

It created a key pair for me (/home/bob/.ssh/id_rsa_hashbang) however, it insisted on using ~/.ssh/id_rsa.pub  when it verified everything

**Comment by necrophcodr** [comment]
It insists on using the other files, because it has no idea you renamed it when the application ssh-keygen was ran.

**Comment by bob109** [comment]
I've tried a few other computers with different versions of ssh and different keys. The end result is always the same.  I also tried the fix referenced in https://github.com/hashbang/hashbang.sh/issues/20 Here's the output from that.

curl -d '{"user":"bob109","key":"'"$(cat ~/.ssh/id_rsa.hashbang.pub | awk '{ print $1 " " $2 }')"'"}' -H 'Content-Type: application/json' https://hashbang.sh/user/create
{"message": "User creation script failed"}

I also ran the following to echo the public ssh key (as seen in another issue)
echo '{"user":"bob109","key":"'"$(cat ~/.ssh/id_rsa.hashbang.pub)"'"}'
{"user":"bob109","key":"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDSu3gi5+IeFpINEpWEd9g+jfkm0lvhlJyMhybg7FJEsvP6rdyRxQsIUR69V1jL7tF89gHBNaa1XUDCZPw0UggU7tQehJJSiumOzKcBlBAS2VZlvkWa74AN8mH1J2BHWknsr9j3IyICgRGZi5ImA+9nB9CYZrdL0XINl+rlDqpKsCG9IAD7yu5Gj0oVr/SozZcyLsMqTqlC7eJZ2e1/RnV0PukXI/VIYalE7RM4vgbjeDAocnMk0tknoni1UtTVPm3dNREOvS64m0/KDzs0bfLL+jcyVFS9Jqw0ztgh0Gz7Sh+BabEH5L8d5n/c9IjOp0v5V3HclLS+t3n9S5EDY28R"}

**Comment by dpflug** [comment]
[How paranoid](https://www.schneier.com/blog/archives/2013/09/the_nsa_is_brea.html#c1675929) do we want to be?

**Comment by stets** [comment]
Receiving same error on two different systems.

**Comment by lrvick** [comment]
Can you guys give it another shot? I reworked several things. Should be more verbose/reliable now.

**Comment by bob109** [comment]
Getting closer. 

I re-downloaded it in the event you also changed the client side. 
When I first ran it, I tried to use an ssh key pair I created in advance.  The script said the key didn't exist.

>We found a public key in [ ~/.ssh/id_rsa.pub ]. Use this key? [Y/n] n
>Do you want us to generate a key for you? [Y/n] n
> 
>Please enter path to SSH Public Key: 
>~/.ssh/id_rsa.hashbang.pub
> 
 >"~/.ssh/id_rsa.hashbang.pub" does not exist
 
I re-ran it and allowed it to use my existing ~/.ssh/id_rsa.pub.  This time it went completely through.  There is a minor problem with the permissions on the ~/.ssh/config file created.  It was created with permissions at 664.  When trying to connect I got the message

>Bad owner or permissions on /home/bob/.ssh/config

I did a chmod 600 ~/.ssh/config and it's all good now.. Thanks a bunch! We can probably consider this issue closed.  I'll submit a new one for the permissions problem if you want to keep it separate. 

bob
